erts/etc/unix/start.src

```diff
@@ -27,14 +27,54 @@
 #
 if [ -z "$ROOTDIR" ]
 then
-   ROOTDIR=%FINAL_ROOTDIR%
+    ROOTDIR=%FINAL_ROOTDIR%
 fi
 
 if [ -z "$RELDIR" ]
 then
-   RELDIR=$ROOTDIR/releases
+    RELDIR=$ROOTDIR/releases
+fi
+
+# If ROOTDIR is not set or if it doesn't point to an existing directory, try
+# to find the path of the script from within itself to set ROOTDIR dynamically
+if [ -z "$ROOTDIR" -o ! -d "$ROOTDIR" ]
+then
+    prog="$0"
+    # Follow all symlinks, whether relative or absolute, leading to the script
+    while [ -h "${prog}" ]; do
+        newProg=`ls -ld "${prog}"`
+        newProg=`expr "${newProg}" : ".* -> \(.*\)$"`
+        if expr "x${newProg}" : 'x/' >/dev/null; then
```

Comment:
I guess this checks if $newProg is an absolute path but I don't fully understand how it works. Can you explain how `expr "x${newProg}" : 'x/' >/dev/null;` works?

Reply:
Hi Kjell,

Indeed, this checks if `newProg` starts with a leading `/` character via a regular expression match:
```
$ expr "x/test" : 'x/' > /dev/null
$ echo $?
0

$ expr "xtest" : 'x/' > /dev/null
$ echo $?
1
```
=> the expression returns 0 if there is a match, 1 otherwise.


P.S. The returned output is the number of characters that matched, but this is not relevent hence the `>/dev/null` part
```
$ expr "x/test" : 'x/'
2
```
In the above case, it shows that the first 2 characters of "x/test" have matched with 'x/'

Reply:
Thank you for the explanation! The x character confused me.

Is the following:

    expr "x${newProg}" : 'x/' > /dev/null"

not equivalent to:

    expr "${newProg}" : '/' > /dev/null"

?

Reply:
They are not equivalent in all cases depending on the value of `newProg`. You can see a similar change here:  

> Be more POSIXly correct in the use of expr.
> https://android.googlesource.com/platform/dalvik/+/530e0a50c365bd8451209a014fef4cf293e573c1

that uses exactly the same convention. I won't pretend to be a script guru to explain it in details though.

You can see an example suggesting to use `expr X$a = X=` reliably here as another reference:
https://pubs.opengroup.org/onlinepubs/009695399/utilities/expr.html

Reply:
I see, x is used to make sure "${newProg}" is not interpreted as an operation. I think it would be good to add some comments about the non-obvious parts to make it more readable by people that are not so familiar with shell scripting.

Btw, I have got a question about if one could use a command like `realpath` instead of doing the link traversal in the script. I will investigate `realpath` but please let me know if you have some insight into that.

Reply:
Hi @kjellwinblad ,

We can't use the `realpath` command as it is not always available on Unix-like systems. In particular, it is not offered natively on Mac OS, see for instance: https://github.com/ethereumjs/ethereumjs-config/issues/41

I'll add a few more comments about the non-obvious script parts, sure.

Let me know when you plan to merge your updated commit discussed at https://github.com/erlang/otp/pull/2863 on master, so that I can rebase on it. Your latest version https://github.com/kjellwinblad/otp/commit/a1259771939d072b2170bb12d3dd3b1fe329875d looks good to me by the way.

Reply:
Let me know what you think about the comments I've just added !

Reply:
Thanks for the explanation about realpath.

I think the comments are good. They make it much easier for readers to understand what is happening.

@rickard-green  suggested that we could make our own version of realpath (based on the C function [realpath](https://pubs.opengroup.org/onlinepubs/009695399/functions/realpath.html)) that we can put under the same directory as the erl script. The [realpath](https://pubs.opengroup.org/onlinepubs/009695399/functions/realpath.html) C function is part of a POSIX standard so it is hopefully available on all relevant platforms. The main benefits of this approach as I see it is that it is potentially more robust and that we would need less complex code in the erl and start scripts. @JeromeDeBretagne What do you think about this idea?

Reply:
Since this scripting has been used in the Android Open Source Project for many years, I believe this is a good sign that it is quite rubust. I usually prefer incremental changes, and the current proposal is a limited modification of existing scripts that have proven to support many cases (native, cross, embedeed, etc.) with only an added dependency on `ls`. 

I'm not a fan at all of the idea to add another executable in the Erlang installation directory, needed to simply launch `erl`, I don't see that as adding robustness. If the Erlang/OTP team is considering bigger changes, I would suggest modifications that vastly **simplify** the Erlang runtime build and installation, for instance :
- decrease the number of installed executables
- get rid of some scripts in favor of native executables
- remove duplicates between `bin/` and `erts-$VSN$/bin`
- remove the `./Install` step
- better support for incrementable builds

That's only my (very personal) view though and it could be discussed within one of the EEF Working Groups I imagine.

@kjellwinblad @rickard-green I still plan to spend more time investigating how to add relevant test cases as suggested, so please let me know if the team still considers this PR a good direction. Thank you.

Reply:
@JeromeDeBretagne We have discussed this pull-request a lot internally in the OTP team today. I think you can take a break from working with this until we have decided what approach to take.

There is already a C program call [dyn_erl.c](https://github.com/erlang/otp/blob/master/erts/etc/unix/dyn_erl.c) that gets installed in bin directory of a release and seems to do similar things as your script. This program was created a long time ago (I did not know about it until today). So other good solutions might be to call that program if $ROOTDIR is incorrect or replace the script with that program. We will let you know when we have made a decision. Thank you for your work so far and sorry if I made you do something that might not be accepted.

Reply:
Thank you @kjellwinblad for your feedback and no worries at all, I've learnt a lot while working on this contribution.

I didn't know about `dyn_erl` until today either. It doesn't seem covered much in the official documentation and it's not found in the main `bin` directory (but it's in the `erts-$VSN%/bin`). That's what I meant by "simplify" in my previous comment :-) 

Good catch, `dyn_erl` looks very similar to what I was trying to achieve indeed and there is already a "realpath look alike" method commented out. I'm really keen to see which direction the OTP team will take on this topic!

Reply:
Sorry for jumping in, but @hawk   do you remember anything about this?
I guess (without looking) that this is your work a long long time ago,
you had the same ideas and did it for windows, why was it never done for unix?
Any lessons you learned and remember that we don't?  :-)